### PR TITLE
KON-248 Refactor With Without Sequence Extensions

### PR DESCRIPTION
**Pull request by nataliapeterwas**

### CODE REVIEW

File: lib/src/main/kotlin/com/lemonappdev/konsist/api/declaration/KoDeclaration.kt [hunk 1]
```diff
@@ -15,7 +15,7 @@ interface KoDeclaration : KoNamedDeclaration {
     /**
      * Package name of the declaration.
      */
```
Comment (nataliapeterwas) [inline lib/src/main/kotlin/com/lemonappdev/konsist/api/declaration/KoDeclaration.kt:17]: I changed logic of the declaration package - but I still not sure if it makes sense. We will discuss it.

**Comment by nataliapeterwas** [comment]
Pay attention to the KDocs - I'm not sure these descriptions are correct :D

File: lib/src/main/kotlin/com/lemonappdev/konsist/api/ext/sequence/KoClassDeclarationSequenceExt.kt [hunk 2]
```diff
@@ -186,16 +186,21 @@ fun Sequence<KoClassDeclaration>.withSecondaryConstructors(): Sequence<KoClassDe
 fun Sequence<KoClassDeclaration>.withoutSecondaryConstructors(): Sequence<KoClassDeclaration> = filterNot { it.hasSecondaryConstructors() }
 
 /**
- * Sequence containing all classes that have parents.
+ * Sequence containing all classes that have any parent.
```
Comment (igorwojda) [inline lib/src/main/kotlin/com/lemonappdev/konsist/api/ext/sequence/KoClassDeclarationSequenceExt.kt:189]: ...have class or interface parent.

Comment (igorwojda) [inline lib/src/main/kotlin/com/lemonappdev/konsist/api/ext/sequence/KoClassDeclarationSequenceExt.kt:189]: ...have class or interface parent.

File: lib/src/main/kotlin/com/lemonappdev/konsist/api/ext/sequence/KoClassDeclarationSequenceExt.kt [hunk 3]
```diff
@@ -210,15 +215,35 @@ fun Sequence<KoClassDeclaration>.withSomeParents(name: String, vararg names: Str
 }
 
 /**
- * Sequence containing all classes that don't have parents.
+ * Sequence containing all classes that have no parent.
+ *
+ * @return A sequence containing classes that have no parent.
```
Comment (igorwojda) [inline lib/src/main/kotlin/com/lemonappdev/konsist/api/ext/sequence/KoClassDeclarationSequenceExt.kt:220]: ..no parent - class does not extend any class and does not implement any interface.

File: lib/src/main/kotlin/com/lemonappdev/konsist/api/ext/sequence/KoClassDeclarationSequenceExt.kt [hunk 4]
```diff
@@ -210,15 +215,35 @@ fun Sequence<KoClassDeclaration>.withSomeParents(name: String, vararg names: Str
...
+ * @param names The names of the parents to exclude.
+ * @return A sequence containing classes that don't have at least one of the specified parent(s).
+ */
+fun Sequence<KoClassDeclaration>.withoutSomeParents(name: String, vararg names: String): Sequence<KoClassDeclaration> = filter {
+    !it.hasParents(name) && if (names.isNotEmpty()) {
+        names.any { name -> !it.hasParents(name) }
+    } else {
+        true
     }
 }
```
Comment (igorwojda) [inline lib/src/main/kotlin/com/lemonappdev/konsist/api/ext/sequence/KoClassDeclarationSequenceExt.kt:249]: IN this method and all methods that have `name` or `names` arguments change "parent of type" to "names parents".

Same for "parent interface of type" -> "named parent interface"

**Review by igorwojda — CHANGES_REQUESTED** [review]

**Review by igorwojda — APPROVED** [review]

**Pull request merged**